lib/common/uint128.h

```diff
@@ -0,0 +1,273 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Portable 128-bit unsigned integer arithmetic for use in impoverished
+// C++ implementations lacking __uint128_t.
+
+#ifndef FORTRAN_COMMON_UINT128_H_
+#define FORTRAN_COMMON_UINT128_H_
+
+#ifndef AVOID_NATIVE_UINT128_T
+#define AVOID_NATIVE_UINT128_T 1  // always use this code for now for testing
+#endif
+
+#include "leading-zero-bit-count.h"
+#include <cstdint>
+#include <type_traits>
+
+namespace Fortran::common {
+
+class UnsignedInt128 {
+public:
+  constexpr UnsignedInt128() {}
+  constexpr UnsignedInt128(std::uint64_t n) : low_{n} {}
+  constexpr UnsignedInt128(std::int64_t n)
+    : low_{static_cast<std::uint64_t>(n)}, high_{-static_cast<std::uint64_t>(
+                                               n < 0)} {}
+  constexpr UnsignedInt128(int n)
+    : low_{static_cast<std::uint64_t>(n)}, high_{-static_cast<std::uint64_t>(
+                                               n < 0)} {}
+  constexpr UnsignedInt128(const UnsignedInt128 &) = default;
+  constexpr UnsignedInt128(UnsignedInt128 &&) = default;
+  constexpr UnsignedInt128 &operator=(const UnsignedInt128 &) = default;
+  constexpr UnsignedInt128 &operator=(UnsignedInt128 &&) = default;
+
+  constexpr UnsignedInt128 operator+() const { return *this; }
+  constexpr UnsignedInt128 operator~() const { return {~high_, ~low_}; }
+  constexpr UnsignedInt128 operator-() const { return ~*this + 1; }
+  constexpr bool operator!() const { return !low_ && !high_; }
+  constexpr explicit operator bool() const { return low_ || high_; }
+  constexpr explicit operator std::uint64_t() const { return low_; }
+  constexpr explicit operator int() const { return static_cast<int>(low_); }
+
+  constexpr std::uint64_t high() const { return high_; }
+  constexpr std::uint64_t low() const { return low_; }
+
+  constexpr UnsignedInt128 operator++(/*prefix*/) {
+    *this += 1;
+    return *this;
+  }
+  constexpr UnsignedInt128 operator++(int /*postfix*/) {
+    UnsignedInt128 result{*this};
+    *this += 1;
+    return result;
+  }
+  constexpr UnsignedInt128 operator--(/*prefix*/) {
+    *this -= 1;
+    return *this;
+  }
+  constexpr UnsignedInt128 operator--(int /*postfix*/) {
+    UnsignedInt128 result{*this};
+    *this -= 1;
+    return result;
+  }
+
+  constexpr UnsignedInt128 operator&(UnsignedInt128 that) const {
+    return {high_ & that.high_, low_ & that.low_};
+  }
+  constexpr UnsignedInt128 operator|(UnsignedInt128 that) const {
+    return {high_ | that.high_, low_ | that.low_};
+  }
+  constexpr UnsignedInt128 operator^(UnsignedInt128 that) const {
+    return {high_ ^ that.high_, low_ ^ that.low_};
+  }
+
+  constexpr UnsignedInt128 operator<<(UnsignedInt128 that) const {
+    if (that >= 128) {
+      return {};
+    } else if (that == 0) {
+      return *this;
```

Comment:
The shift operators could be simplified by replacing the first two special cases with `if (that.high_ > 0) return {}`.